### PR TITLE
[withStyles] Forward ref to inner component

### PR DESCRIPTION
**Pull request by indiesquidge**
Refs will not get passed through withStyles by default because `ref` is not a prop. If you add a ref to an HOC, the ref will refer to the outermost container component, not the wrapped component.

Ref: https://reactjs.org/docs/forwarding-refs.html#forwarding-refs-in-higher-order-components

### CODE REVIEW

**Comment by indiesquidge** [comment]
There are 3 failing tests that I'm not quite sure how to fix.
```
1) withStyles(styleFn)(Component) sets the displayName:
   AssertionError: expected undefined to equal 'withStyles(MockComponent)'

2) withStyles(styleFn)(Component) copies the component's propTypes and defaultProps, except the withStyles ones:
   TypeError: Cannot convert undefined or null to object

3) withStyles Without providers in a non-directional context uses the theme registered with the singleton API:
   AssertionError: expected false to equal true
```

I believe they're related to hoisting non react statics, but I can't figure out what I'm doing wrong. I've followed the [React docs][1] and [`hoist-non-react-statics` implementations][2] as I understand them.

[1]: https://reactjs.org/docs/forwarding-refs.html#forwarding-refs-in-higher-order-components
[2]: https://github.com/mridgway/hoist-non-react-statics/blob/602ca35b5b3054f25164a396d71be1d25e0c007d/tests/unit/index.js#L200-L254

File: src/withStyles.jsx [hunk 1]
```diff
@@ -228,20 +231,27 @@ export function withStyles(
       }
     }
 
+    const ForwardedWithStyles = React.forwardRef((props, ref) => (
```
Comment (lencioni) [inline src/withStyles.jsx]: I think forwardRef was added in 16.3, but react-with-styles currently works all the way down to 0.14 https://github.com/airbnb/react-with-styles/blob/59a9548c7d18d8bf8bcf8a99cea85548114edf73/package.json#L81

Do you think it is possible to do this in a backwards compatible way, e.g. by detecting the presence of `React.forwardRef`

Comment (ljharb) [reply src/withStyles.jsx]: The only thing I can think of would be adding an alternative entry point, and leave this one as-is - forwardRef doesn't really have any alternative in React < 16.3 that I'm aware of.

Comment (lencioni) [reply src/withStyles.jsx]: Would it work to do something like this? (along with a guard in the render function so it conditionally passes it down to the wrapped component).

```js
const ForwardedWithStyles = typeof React.forwardRef === 'undefined'
  ? WithStyles
  : <WithStyles {...props} forwardedRef={ref} />;
```

That way, older versions continue to work as they do today, and 16.3+ will have the benefits of forwardRef.

Comment (ljharb) [reply src/withStyles.jsx]: The `forwardedRef` propType (which i don't see added anywhere yet) would need to be conditionally applied, but that might work - altho it seems a bit weird to me to have an entire capability only be conditionally present based on react version.

File: test/withStyles_test.jsx [hunk 2]
```diff
@@ -570,7 +570,7 @@ describe('withStyles', () => {
         const MockComponent = ({ styles, css }) => <div {...css(styles.primary)} />;
         MockComponent.propTypes = { ...withStylesPropTypes };
         const StyledComponent = withStyles(stylesFn)(MockComponent);
-        const wrapper = shallow(<StyledComponent />);
+        const wrapper = mount(<StyledComponent />);
```
Comment (ljharb) [inline test/withStyles_test.jsx]: why is this change needed?

File: src/withStyles.jsx (hunk 1 above)
Comment (lencioni) [reply src/withStyles.jsx]: Yeah that's fair. Might as well make a breaking change. 16.3 was released almost 2 years ago now.

File: test/withStyles_test.jsx (hunk 2 above)
Comment (indiesquidge) [reply test/withStyles_test.jsx]: Not super sure actually. I commented about it here https://github.com/airbnb/react-with-styles/pull/240#issuecomment-589309710

File: src/withStyles.jsx (hunk 1 above)
Comment (indiesquidge) [reply src/withStyles.jsx]: > The forwardedRef propType (which i don't see added anywhere yet) would need to be conditionally applied

Can you explain what do you mean you don't see it added anywhere yet?

Comment (ljharb) [reply src/withStyles.jsx]: An explicit propType that declares `forwardedRef` on `WithStyles`

File: test/withStyles_test.jsx (hunk 2 above)
Comment (ljharb) [reply test/withStyles_test.jsx]: ah - you'd need to add a `.dive()` here to invoke the deeper level (one dive per HOC/wrapper)

File: src/withStyles.jsx (hunk 1 above)
Comment (indiesquidge) [reply src/withStyles.jsx]: > An explicit propType that declares forwardedRef on WithStyles

Should this exist though? It should only be added by implicitly by `React.forwardRef`. A consumer of `withStyles` would not need to know about this prop. I don't think I'm understanding the requested change.

Comment (indiesquidge) [reply src/withStyles.jsx]: Are we looking to have [something like this](https://stackoverflow.com/a/51127130)?

```jsx
const withStylesPropTypes = {
  refProp: PropTypes.oneOfType([
      PropTypes.func, 
      PropTypes.shape({ current: PropTypes.any })
  ])
}
```

Comment (ljharb) [reply src/withStyles.jsx]: You could use `ref` from `airbnb-prop-types` :-p

Comment (indiesquidge) [reply src/withStyles.jsx]: Nice. I added this change

```jsx
```

**Comment by ljharb** [comment]
If we're going with the breaking change, we'll need to update the peer deps to require react 16.3+; if not (my preference), probably best to add a second entry point that contains these changes, and leave the existing one intact.

**Comment by yoution** [comment]
when will this new function  be merged to master branch？

File: src/withStyles.jsx (hunk 1 above)
Comment (indiesquidge) [reply src/withStyles.jsx]: @ljharb @lencioni what're our current thoughts here? Should we [conditionally fork][1] the existing entry point, or create a [brand new][2] entry point? Or create a breaking change given the release age of 16.3 now?

[1]: https://github.com/airbnb/react-with-styles/pull/240#discussion_r382745271
[2]: https://github.com/airbnb/react-with-styles/pull/240#discussion_r382740336

Comment (lencioni) [reply src/withStyles.jsx]: I think it makes sense to publish a version with conditionally forking, and then immediately follow up with a breaking change to remove the conditional fork.

Comment (ljharb) [reply src/withStyles.jsx]: Makes sense to me.

File: src/withStyles.jsx [hunk 3]
```diff
@@ -249,9 +251,21 @@ export function withStyles(
...
+      };
+    }
+
+    // TODO: remove conditional once breaking change to only support React 16.3+
+    // ref: https://github.com/airbnb/react-with-styles/pull/240#discussion_r533497857
+    const ForwardedWithStyles = typeof React.forwardRef === 'undefined'
+      ? WithStyles
+      : React.forwardRef((props, forwardedRef) => (
+        <WithStyles {...props} forwardedRef={forwardedRef} />
+      ));
```
Comment (indiesquidge) [inline src/withStyles.jsx:268]: @lencioni @ljharb is this commit what you had in mind for the forking?

**Review by ljharb — APPROVED** [review]
Seems reasonable to me.

File: src/withStyles.jsx [hunk 4]
```diff
@@ -241,20 +247,43 @@ export function withStyles(
...
+      forwardedRef: ref,
+    };
+
+    // TODO: remove conditional once breaking change to only support React 16.3+
+    // ref: https://github.com/airbnb/react-with-styles/pull/240#discussion_r533497857
+    if (typeof React.forwardRef !== 'undefined') {
+      WithStyles.propTypes = {
+        forwardedRef: ref,
+      };
+    }
```
Comment (lencioni) [inline src/withStyles.jsx:260]: Is the first propTypes assignment here a mistake?
```suggestion
    // TODO: remove conditional once breaking change to only support React 16.3+
    // ref: https://github.com/airbnb/react-with-styles/pull/240#discussion_r533497857
    if (typeof React.forwardRef !== 'undefined') {
      WithStyles.propTypes = {
        forwardedRef: ref,
      };
    }
```

**Review by lencioni — COMMENTED** [review]
Will you please add a test and update documentation?

File: src/withStyles.jsx (hunk 4 above)
Comment (indiesquidge) [reply src/withStyles.jsx:260]: Good catch! That diff must've snuck in during a rebase.

File: test/withStyles_test.jsx [hunk 5]
```diff
@@ -142,6 +142,18 @@ describe('withStyles', () => {
...
+      const MockComponent = React.forwardRef((props, forwardedRef) => (
+        <div ref={forwardedRef} />
+      ));
+      const StyledComponent = withStyles(stylesFn)(MockComponent);
+      const ref = React.createRef();
+      const wrapper = mountWithProviders(<StyledComponent ref={ref} />, providers);
+      const div = wrapper.find('div').first().getDOMNode();
+      expect(div).to.equal(ref.current);
+    });
+
```
Comment (indiesquidge) [inline test/withStyles_test.jsx:156]: I didn't include an inverse of this test to assert that, for React<16.3, refs are _not_ forwarded because:

- I wasn't sure how to set up a test that runs for React<16.3
- I didn't think the test would be necessary if we plan on adding a breaking change to drop support for React<16.3 after this PR is merged

File: README.md [hunk 6]
```diff
@@ -420,6 +420,41 @@ export default withStyles(({ color, unit }) => ({
...
+    color: color.primary,
+    marginBottom: 2 * unit,
+  },
+}))(MyComponent);
+
+// the ref will be passed down to MyComponent, which is then attached to the div
+const ref = React.createRef()
+<MyComponentWithStyles ref={ref} />
+```
+
```
Comment (indiesquidge) [inline README.md:457]: @ljharb @lencioni let me know if this documentation addition sounds good to you.
